qemu/tests/blockdev_stream_backing_mask_on.py

```diff
@@ -0,0 +1,97 @@
+import logging
+
+from virttest import data_dir
+
+from provider import backup_utils
+from provider.blockdev_stream_base import BlockDevStreamTest
+from provider.virt_storage.storage_admin import sp_admin
+
+LOG_JOB = logging.getLogger('avocado.test')
+
+
+class BlockdevStreamBackingMaskOnTest(BlockDevStreamTest):
+    """Do block-stream based on an existed snapshot in snapshot chain"""
+
+    def __init__(self, test, params, env):
+        super(BlockdevStreamBackingMaskOnTest, self).__init__(test, params, env)
+        self._snapshot_images = self.params.objects("snapshot_images")
+        backing_mask_protocol = self.params["backing_mask_protocol"]
+        if backing_mask_protocol:
+            self._stream_options["backing-mask-protocol"] = backing_mask_protocol
+        self._trash = []
+
+    def snapshot_test(self):
+        """create one snapshot, create one new file"""
+        self.generate_tempfile(self.disks_info[self.base_tag][1],
+                               filename="base",
+                               size=self.params["tempfile_size"])
+
+        # data->sn1->sn2->sn3->sn4
+        chain = [self.base_tag] + self._snapshot_images
+        for idx in range(1, len(chain)):
+            backup_utils.blockdev_snapshot(
+                self.main_vm,
+                "drive_%s" % chain[idx-1],
+                "drive_%s" % chain[idx]
+            )
+
+            self.generate_tempfile(self.disks_info[self.base_tag][1],
+                                   filename=chain[idx],
+                                   size=self.params["tempfile_size"])
+
+    def _disk_define_by_params(self, tag):
+        params = self.params.copy()
+        params.setdefault("target_path", data_dir.get_data_dir())
+        return sp_admin.volume_define_by_params(tag, params)
+
+    def prepare_snapshot_file(self):
+        """hotplug all snapshot images"""
+        for tag in self._snapshot_images:
+            disk = self._disk_define_by_params(tag)
+            disk.hotplug(self.main_vm)
+            self._trash.append(disk)
+
+    def _remove_images(self):
+        for img in self._trash:
+            sp_admin.remove_volume(img)
+
+    def post_test(self):
+        try:
+            if self.main_vm.is_alive():
+                self.main_vm.destroy()
+            self._remove_images()
+        except Exception as e:
+            LOG_JOB.warning(str(e))
+
+    def check_backing_format(self):
+        base_format = self.base_image.get_format()
+        output = self.snapshot_image.info(force_share=True).split("\n")
+        for item in output:
+            if "backing file format" in item:
```

Comment:
Not a comment, but a minor reminder. Keep an eye on the  .. else... , e.g. what if there is never a 'backing file format' in the output, our test also passed, but the fact is line71 is never checked.

Reply:
please also check https://github.com/autotest/tp-qemu/pull/4141